src/lib.rs

```diff
@@ -74,7 +76,7 @@ where
         let self_ = self.clone();
         let query = query.to_string();
         task::block_in_place(move || {
-            let conn = self_.get().map_err(AsyncError::Checkout)?;
+            let conn = self_.try_get().ok_or(AsyncError::Checkout)?;
```

Comment:
I don't think we'd want to universally prevent waiting for a connection when the pool is fully utilized. You can get very similar behavior by setting the r2d2 connection_timeout builder option to a very small positive duration.

Reply:
If we want to wait, let's do it asynchronously as otherwise, it is blocking the current thread (`select!` and `join!` on multiple async actions does not progress)

Reply:
Wouldn't the `task::block_in_place` allow other futures to continue polling since `block_in_place` uses a dedicated threadpool for blocking functions?

In any case, it would be nice to build a custom future to non-block get a connection. However, if you simply do `try_get` in a future over and over, you could end up starved for a connection since order would not be guaranteed.

Reply:
I believe I spoke incorrectly about the behavior of `block_in_place`. I was thinking of `spawn_blocking` and not the recent change to `block_in_place`.

Reply:
Yep, that was `spawn_blocking`, but with the receipt change to [`block_in_place` we hit this documented behavior](https://docs.rs/tokio/0.2.22/tokio/task/fn.block_in_place.html):

> Although this function avoids starving other independently spawned tasks, any other code running concurrently in the same task will be suspended during the call to block_in_place. This can happen e.g. when using the join! macro. To avoid this issue, use spawn_blocking instead.